inference-engine/include/ie_extension.h

```diff
@@ -126,4 +126,77 @@ inline std::shared_ptr<IExtension> make_so_pointer(const std::wstring& name) {
 }
 
 #endif
+
+// ===========================================================================================================================================
+// New extensions experiment
+// ===========================================================================================================================================
+class INFERENCE_ENGINE_API_CLASS(NewExtension) {
+public:
+    using Ptr = std::shared_ptr<NewExtension>;
+};
+
+class INFERENCE_ENGINE_API_CLASS(ExtensionContainer) {
+public:
+    using Ptr = std::shared_ptr<ExtensionContainer>;
+    ExtensionContainer() = default;
+    ExtensionContainer(const std::vector<NewExtension::Ptr>& extensions): extensions(extensions) {} // NOLINT
+
+    virtual ~ExtensionContainer() = default;
+    virtual const std::vector<NewExtension::Ptr>& getExtensions() const {
+        return extensions;
+    }
+    operator const std::vector<NewExtension::Ptr>() const {
+        return getExtensions();
+    }
+
+protected:
+    std::vector<NewExtension::Ptr> extensions;
+};
+
+namespace details {
+
+/**
+ * @brief The SOCreatorTrait class specialization for IExtension case, defines the name of the fabric method for
+ * creating IExtension object in DLL
+ */
+template <>
+class SOCreatorTrait<ExtensionContainer> {
+public:
+    /**
+     * @brief A name of the fabric method for creating an IExtension object in DLL
+     */
+    static constexpr auto name = "CreateExtensionContainer";
+};
+
+}  // namespace details
+
+class INFERENCE_ENGINE_API_CLASS(SOExtension) final: public ExtensionContainer {
+public:
+    template <typename C,
+              typename = details::enableIfSupportedChar<C>>
+    explicit SOExtension(const std::basic_string<C>& name): actual(name) {}
+
+
+    const std::vector<NewExtension::Ptr>& getExtensions() const override {
+        return actual->getExtensions();
+    }
+private:
+    details::SOPointer<ExtensionContainer> actual;
+};
+
+/**
+ * @brief Creates the default instance of the extension
+ *
+ * @param ext Extension interface
+ */
+INFERENCE_EXTENSION_API(void) CreateExtensionContainer(ExtensionContainer::Ptr& ext);
+
+/**
+ * @def IE_DEFINE_EXTENSION_CREATE_FUNCTION
+ * @brief Generates extension creation function
+ */
+#define IE_CREATE_EXTENSION(ExtensionType)                                                                                              \
+INFERENCE_EXTENSION_API(void) InferenceEngine::CreateExtensionContainer(std::shared_ptr<InferenceEngine::ExtensionContainer>& ext) {    \
+    ext = std::make_shared<ExtensionType>();                                                                                            \
```

Comment:
It means you have to derive a new class from `ExtensionContainer` to populate it with multiple items instead of just create an instance of `Extensioncontainer` with necessary items via vector. It looks overcomplicated. Could we pass a list of desired extensions in the macro arguments and construct `ExtensionContainer` with those arguments in a vector.